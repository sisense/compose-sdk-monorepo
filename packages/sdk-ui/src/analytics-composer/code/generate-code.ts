import { removeEmptyLineBeforeImport } from '../common/utils.js';
import { CodePlaceholderMap, CodeTemplateKey, UiFramework } from '../types.js';
import { CODE_TEMPLATES } from './code-templates.js';

/**
 * Populates templates recursively.
 *
 * @param templateKey - the key of the template to use
 * @param uiFramework - the UI framework for the generated code
 * @returns the populated template with placeholders
 */
const populateTemplates = (templateKey: CodeTemplateKey, uiFramework: UiFramework): string => {
  const template = CODE_TEMPLATES[uiFramework][templateKey];
  return (
    template
      // replace template placeholders with actual templates recursively
      .replace(/{{(.*?)Tmpl}}/g, (match, key) =>
        populateTemplates(`${key}Tmpl` as CodeTemplateKey, uiFramework),
      )
  );
};

/**
 * Populates placeholders in the given template with the given data.
 * If a placeholder is not found in the data, a warning is logged.
 *
 * @param template - the template to populate
 * @param data - the data to populate the template with
 * @returns the populated template
 */
export const populatePlaceholders = (template: string, data: Record<string, string>): string => {
  // replace other placeholders with actual values defined in data
  return template.replace(/{{(.*?)}}/g, (match, key) => {
    if (key in data) {
      return data[key];
    }
    console.warn(`Placeholder ${key} not found in data`);
    return match;
  });
};

/**
 * Generates code from a template with the given data and UI framework.
 *
 * @param templateKey - the key of the template to use
 * @param placeholderMap - the data to populate the template with
 * @param uiFramework - the UI framework for the generated code
 */
export const generateCode = (
  templateKey: CodeTemplateKey,
  placeholderMap: CodePlaceholderMap,
  uiFramework: UiFramework = 'react',
): string => {
  const template = populateTemplates(templateKey, uiFramework);
  return removeEmptyLineBeforeImport(populatePlaceholders(template, placeholderMap));
};
