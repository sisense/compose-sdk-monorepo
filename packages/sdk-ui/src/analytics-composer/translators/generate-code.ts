import { CODE_TEMPLATES } from './code-templates';
import { CodeTemplateKey, UiFramework } from './types';

/**
 * Populates templates recursively.
 *
 * @param templateKey - the key of the template to use
 * @param uiFramework - the UI framework for the generated code
 * @returns the populated template with placeholders
 */
const populateTemplates = (templateKey: CodeTemplateKey, uiFramework: UiFramework): string => {
  const template = CODE_TEMPLATES[uiFramework][templateKey];
  return (
    template
      // replace template placeholders with actual templates recursively
      .replace(/{{(.*?)Tmpl}}/g, (match, key) =>
        populateTemplates(`${key}Tmpl` as CodeTemplateKey, uiFramework),
      )
  );
};

/**
 * Populates placeholders in the given template with the given data.
 * If a placeholder is not found in the data, a warning is logged.
 *
 * @param template - the template to populate
 * @param data - the data to populate the template with
 * @returns the populated template
 */
export const populatePlaceholders = (template: string, data: Record<string, string>): string => {
  // replace other placeholders with actual values defined in data
  return template.replace(/{{(.*?)}}/g, (match, key) => {
    if (key in data) {
      return data[key];
    }
    console.warn(`Placeholder ${key} not found in data`);
    return match;
  });
};

/**
 * Generates code from a template with the given data and UI framework.
 *
 * @param templateKey - the key of the template to use
 * @param data - the data to populate the template with
 * @param uiFramework - the UI framework for the generated code
 */
export const generateCode = (
  templateKey: CodeTemplateKey,
  data: Record<string, string>,
  uiFramework: UiFramework = 'react',
): string => {
  const template = populateTemplates(templateKey, uiFramework);
  return populatePlaceholders(template, data);
};
