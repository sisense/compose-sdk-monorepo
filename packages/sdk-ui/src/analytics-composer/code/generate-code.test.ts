import { generateCode } from './generate-code.js';

const dataForTemplate: Record<string, string> = {
  componentString: 'Chart',
  extraImportsString: `import { filterFactory, measureFactory } from '@sisense/sdk-data';`,
  dataSourceString: 'data-source',
  titleString: 'My Chart',
  chartTypeString: 'bar',
  dataOptionsString:
    '{ category: [DM.Commerce.Date.Years], value: [measureFactory.sum(DM.Commerce.Revenue)] }',
  filtersString: '[filterFactory.greaterThan(DM.Commerce.Revenue, 0)]',
};

describe('generateCode', () => {
  it('should generate code from template for React', () => {
    const code = generateCode('chartWidgetTmpl', dataForTemplate);

    expect(code).toMatchSnapshot();
  });

  it('should generate code from template for Vue', () => {
    const code = generateCode('chartWidgetTmpl', dataForTemplate, 'vue');

    expect(code).toMatchSnapshot();
  });

  it('should generate code from template for Angular', () => {
    const code = generateCode('chartWidgetTmpl', dataForTemplate, 'angular');

    expect(code).toMatchSnapshot();
  });
});
