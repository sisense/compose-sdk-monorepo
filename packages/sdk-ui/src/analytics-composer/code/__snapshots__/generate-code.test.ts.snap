// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateCode > should generate code from template for Angular 1`] = `
"import { Component } from '@angular/core';
import { type ChartDataOptions, type ChartStyleOptions } from '@sisense/sdk-ui-angular';
import { filterFactory, measureFactory } from '@sisense/sdk-data';
import * as DM from './data-source'; // generated with @sisense/sdk-cli

@Component({
    selector: 'code-example',
    template: \`
      <csdk-chart-widget
        [title]="'My Chart'"
        chartType='bar'
        [dataSource]='DM.DataSource'
        [dataOptions]='dataOptions'
        [filters]='filters'
        [styleOptions]='styleOptions'
        [drilldownOptions]='drilldownOptions'
      />
    \`
})
export class CodeExample {
    DM = DM;
    dataOptions: ChartDataOptions = { category: [DM.Commerce.Date.Years], value: [measureFactory.sum(DM.Commerce.Revenue)] };
    filters = [filterFactory.greaterThan(DM.Commerce.Revenue, 0)];
    styleOptions: ChartStyleOptions = {{styleOptionsString}};
    drilldownOptions = {{drilldownOptionsString}};
}"
`;

exports[`generateCode > should generate code from template for React 1`] = `
"import { Chart } from '@sisense/sdk-ui';
import { filterFactory, measureFactory } from '@sisense/sdk-data';
import * as DM from './data-source'; // generated with @sisense/sdk-cli

export default function CodeExample() {
  return (
    <Chart
      title={'My Chart'}
      dataSource={DM.DataSource}
      chartType={'bar'}
      dataOptions={ { category: [DM.Commerce.Date.Years], value: [measureFactory.sum(DM.Commerce.Revenue)] } }
      filters={ [filterFactory.greaterThan(DM.Commerce.Revenue, 0)] }
      styleOptions={ {{styleOptionsString}} }
      drilldownOptions={ {{drilldownOptionsString}} }
    />
  );
}"
`;

exports[`generateCode > should generate code from template for Vue 1`] = `
"<script setup lang="ts">
  import { ref } from 'vue';
  import { Chart, type ChartStyleOptions } from '@sisense/sdk-ui-vue';
  import { filterFactory, measureFactory } from '@sisense/sdk-data';
  import * as DM from './data-source'; // generated with @sisense/sdk-cli

  const dataOptions = ref({ category: [DM.Commerce.Date.Years], value: [measureFactory.sum(DM.Commerce.Revenue)] });
  const filters = ref([filterFactory.greaterThan(DM.Commerce.Revenue, 0)]);
  const styleOptions = ref<ChartStyleOptions>({{styleOptionsString}});
  const drilldownOptions = ref({{drilldownOptionsString}});
</script>

<template>
  <Chart
    chartType="bar"
    :dataOptions="dataOptions"
    :filters="filters"
    :styleOptions="styleOptions"
    :drilldownOptions="drilldownOptions"
    :dataSource="DM.DataSource"
    title="My Chart" />
</template>"
`;
